extensions/cms/src/cmsUtils.ts

```diff
@@ -70,49 +104,70 @@ export class CmsUtils {
+		if (result && result.registeredServersList && result.registeredServersList) {
+			return result;
+		}
 	}
 
 	public async createCmsServer(connection: azdata.connection.Connection,
-		name: string, description: string): Promise<mssql.ListRegisteredServersResult> {
+		name: string, description: string): Promise<CreateCmsResult> {
 		let provider = await this.getCmsService();
 		connection.providerName = connection.providerName === cmsProvider ? mssqlProvider : connection.providerName;
 		let ownerUri = await azdata.connection.getUriForConnection(connection.connectionId);
 		if (!ownerUri) {
 			// Make a connection if it's not already connected
-			await azdata.connection.connect(Utils.toConnectionProfile(connection), false, false).then(async (result) => {
-				ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
-			});
+			let initialConnectionProfile = CmsUtils.getConnectionProfile(connection);
+			let result = await azdata.connection.connect(initialConnectionProfile, false, false);
+			ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+			// If the ownerUri is still undefined, then open a connection dialog with the connection
+			if (!ownerUri) {
+				try {
+					let result = await this.connection(initialConnectionProfile);
+					if (result) {
+						ownerUri = await azdata.connection.getUriForConnection(result.connectionId);
+						connection = result;
+					}
+				} catch (error) {
+					// cancel pressed on connection dialog
+					const errorText = localize('cms.error.cancelConnectionDialog', 'The server is disconnected. Please connect to the server to continue.');
+					return Promise.reject(error ? new Error(error.message) : errorText);
+				}
+			}
 		}
-		return provider.createCmsServer(name, description, connection, ownerUri).then((result) => {
+		try {
+			let result = await provider.createCmsServer(name, description, connection, ownerUri);
 			if (result) {
-				return Promise.resolve(result);
-			} else {
-				return Promise.reject(null);
+				const createCmsResult: CreateCmsResult = {
+					listRegisteredServersResult: result,
+					connection: connection
+				};
+				return Promise.resolve(createCmsResult);
 			}
-		});
+		} catch (errorString) {
+			const error = new Error(errorString.message);
+			return Promise.reject(error);
+		}
 	}
 
-	public async deleteCmsServer(cmsServer: any): Promise<void> {
+	public async deleteCmsServer(cmsServerName: string, connection: azdata.connection.Connection): Promise<void> {
 		let config = this.getConfiguration();
 		if (config && config.servers) {
 			let newServers = config.servers.filter((cachedServer) => {
-				return cachedServer.name !== cmsServer;
+				return cachedServer.name !== cmsServerName;
 			});
 			await this.setConfiguration(newServers);
 			this._registeredCmsServers = this._registeredCmsServers.filter((cachedServer) => {
-				return cachedServer.name !== cmsServer;
+				return cachedServer.name !== cmsServerName;
```

Comment:
It seems weird to have two lists of the registered servers - one local and one from the config. Why aren't you just keeping them all in one place and using that alone?